PR for 162: Refactor event details layout for new fix elements

lemonboston opened the pull request:
@dmfs Could you please have a first look on how the `SmartView<Optional<T>>` classes are used now in `EventDetailFragment` along with the layout `include`s and bindings? It looks really useful and there are possibilities to create some more reusable elements there but I don't want to overdo it before you have look.

The task is not completely finished yet, the code part is mostly done, but I need to go through all the looks, colors, sizes, etc, so I will keep commiting to the branch.

Commit:
Refactor event details layout to have the new fixed elements. #162

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/actions/ActionExecutingOnClickListener.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class ActionExecutingOnClickListener implements View.OnClickListener
```
Maybe call it just `ActionClickListener` everything else seems kind of redundant.

lemonboston replied:
Done.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/actions/DirectionsLabelFactory.java:
```diff
@@ -44,6 +44,6 @@ public DirectionsLabelFactory(Iterable<Location> locations)
     public CharSequence create(Context context)
     {
-        return new AddressName(mLocations).value(context.getString(R.string.schedjoules_action_directions));
+        return new AddressName(mLocations).value(context.getString(R.string.schedjoules_action_directions).toUpperCase());
```
Isn't the directions label styled like a BorderlessButton? If possible leave the up-case conversion to the styles.

lemonboston replied:
The Directions action can have 2 labels now: DIRECTIONS in case we have no address and the address in normal case if we have it, although I am not sure if the first case would ever happen.
What would the BorderlessButton style add? Could we still simply add the icon there? Now the whole layout, the icon and the text acts as a button when pressed.
In any case, can we address it in a separate story if needed?

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/actions/OptionalAction.java:
```diff
@@ -0,0 +1,64 @@
+public final class OptionalAction extends AbstractCachingOptional<Action>
+{
+
+    public OptionalAction(final String actionLinkRelType, final Iterable<Link> actionLinks, final ActionFactory actionFactory, final Event event)
```
Why do you pass an `ActionFactory` to this constructor?

lemonboston replied:
There was an `Actions` `Iterable` earlier that created the `Action`s, but I needed to filter them by rel type and instead of adding that as method to `Action` (which I did at one point), I thought it may be better to just filter the `Link`s first and then create that single `Action`. So `OptionalAction` needs the factory to create the `Action` from the corresponding `Link`.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/actions/OptionalAction.java:
```diff
@@ -0,0 +1,64 @@
+                    }
+                });
+                //noinspection ConstantConditions
+                return filteredIterator.hasNext() ?
```
Maybe create a `First` adapter, just like the one in the Optional package. The one in the Optional package is only for `Iterables`, but in this case it would make sense to have one for Iterators too. The downside is that you would have to execute the check eagerly in the constructor (because the Iterator is not immutable).

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/actions/SimpleAction.java:
```diff
@@ -50,7 +51,7 @@ public SimpleAction(CharSequenceFactory labelFactory, @DrawableRes int iconResId
     public SimpleAction(@StringRes int labelResId, @DrawableRes int iconResId, @NonNull ActionExecutable actionExecutable)
     {
-        this(new ResourceCharSequenceFactory(labelResId), iconResId, actionExecutable);
+        this(new UpperCased(new ResourceCharSequenceFactory(labelResId)), iconResId, actionExecutable);
```
Again, the upper case conversion should probably be handled by the style.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/actions/TicketButtonAction.java:
```diff
@@ -0,0 +1,96 @@
+            @Override
+            public Optional<Action> create()
+            {
+                if (!ticketLink.isPresent())
```
I wonder if we can achieve the purpose of this method with a decorator. Something like a "MappedOptional" (similar to the Converted Iterator decorator).
like so:
```java
return new Mapped<Action>(ticketLink, new Function<Link, Action>(){
      public Action mapped(Link link)
      {
        return new InternalTicketAction(link, event);
      }
   });
```

Not sure if we would still need TicketButtonAction in that case.

lemonboston replied:
Created `Mapped` Optional for that and kept `TicketButtonAction` which now hides `BookTicketLink` as well.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/microfragments/eventdetails/fragments/views/ActionView.java:
```diff
@@ -0,0 +1,71 @@
+        }
+        else
+        {
+            mTitleView.setText(null);
```
Is it really necessary to set null values? Isn't is sufficient to set visibility to GONE?

lemonboston replied:
No, it was not, right, I removed them. This still works correctly if it is updated multiple time.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/microfragments/eventdetails/fragments/views/TicketButtonView.java:
```diff
@@ -0,0 +1,65 @@
+    @Override
+    public void update(Optional<Action> bookAction)
+    {
+        mButtonArea.setVisibility(bookAction.isPresent() ? View.VISIBLE : View.GONE);
```
better just call `bookAction.isPresent()` just once and branch with an `if` statement. It might become a bit longer but it's more readable and cleaner.

lemonboston replied:
`OptionalView` is used now to hide that visibility switching.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/widgets/OptionalTextedView.java:
```diff
@@ -0,0 +1,52 @@
+    private final TextView mTextView;
+
+
+    public OptionalTextedView(View root, TextView textView)
```
why does it need both, the View and the Root? Could you pass the view resource id instead of the view itself?

lemonboston replied:
This class has been removed.

dmfs commented on eventdiscovery-sdk/src/main/res/layout/schedjoules_fragment_event_details.xml:
```diff
@@ -24,14 +24,78 @@
+                        android:layout_marginBottom="8dp"/>
+
+                <include
+                        layout="@layout/schedjoules_view_event_details_action"
```
a lot of repetition here. Please create a separate style. Also watch out for the box model. Often it's better to give the element the same margin on all sides (e.g. 8dp) and have the missing 8dp added by a padding on the container element.

lemonboston replied:
Yeah, that was intentional at that point, I wanted to get back to styles and the end, also to see how the layout ends up. I've created a style for these event detail item sizes.
I left the 16dp horizontal padding instead of 8dp and then 8dp on the container, to have a greater area for the touch ripple effect.

dmfs commented on eventdiscovery-sdk/src/main/res/values/dimens.xml:
```diff
@@ -13,7 +13,7 @@
     <dimen name="schedjoules_list_item_default_height">96dp</dimen>
     <dimen name="schedjoules_event_list_progress_bar">36dp</dimen>
-    <dimen name="schedjoules_event_detail_item_icon">28dp</dimen>
+    <dimen name="schedjoules_event_detail_item_icon">24dp</dimen>
```
What's that? Height? Width? Margin? Padding? Spacing? Elevation? … Please add what kind of value this is.

lemonboston replied:
Changed to `schedjoules_event_detail_item_icon_size` as it is both height and width.

dmfs reviewed (changes requested):
Done with the first round.

dmfs commented:
Please also make sure the action items have a padding instead of a margin, otherwise the sensitive area becomes to small and the visual feedback covers only a small area too.

dmfs commented:
The footer doesn't become fully visible any more when you scroll down on a page with expanded description. Also see the attached image.

![device-2017-04-06-000012](https://cloud.githubusercontent.com/assets/1614987/24729063/c82d95ec-1a5c-11e7-915f-fb29d85c5d6b.png)

lemonboston commented:
Changed margins to paddings as suggested, also fixed the bug above, it was caused by the `LinearLayout` in the `NestedScrollView` having top margin, changing to padding fixed it.

lemonboston commented:
Ready for review.

As this changeset got big, I stop commiting to this and branch off for outstanding things that I have in mind:

- Venue name text size is not increased yet, because that causes misalignment of the icon - text. Need to find solution to that.
- Review the event details screen look again compared to design, check all sizes and colors
- Review colors.xml, styles.xml, make any common sense changes, then look through the event details related layouts again to use those or new styles where it make sense. Maybe also create named dimens for text sizes. Check if it make sense to apply primary and secondary text color theme properties. Maybe communicate color palette and usage to ui-ux channel. Ask about sizes on the design (there is a text with 23px for example.). Check divider color (the Android one looked too light).
- BorderlessButton for Actions if needed

I'll start with the first on a separate branch, we can create a bug for that.

lemonboston commented:
There is already a separate ticket for Action labels, we can address any possible change needed there:
https://github.com/schedjoules/android-event-discovery-sdk/issues/168

dmfs reviewed (approved):
Approving this despite concerns regarding the uppercase handling.
Also, make sure to submit a PR that restores the original name of schedjoules_fragment_event_details_action_loading
